docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,210 @@
+
+| Instance Attributes | Type | Description |
+| ------------- | ------------- | ------------- |
+| `niltype` | `typelib` | an immutable runtime representation of the built-in type `nil` |
+| `unknown` | `typelib` | an immutable runtime representation of the built-in type `unknown` |
+| `never` | `typelib` | an immutable runtime representation of the built-in type `never` |
+| `any` | `typelib` | an immutable runtime representation of the built-in type `any` |
+| `boolean` | `typelib` | returns an immutable runtime representation of the built-in type `boolean` |
+| `number` | `typelib` | returns an immutable runtime representation of the built-in type `number` |
+| `string` | `typelib` | returns an immutable runtime representation of the built-in type `string` |
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is syntactically a subtype or equal to arg in the type hierarchy |
+| `__eq(arg: typelib)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg in the type hierarchy |
+
+| Static Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |
+| `getstringsingleton(arg: string)` | `typelib` | returns an immutable runtime representation of a string singleton type of the argument |
+| `getbooleansingleton(arg: boolean)` | `typelib` | returns an immutable runtime representation of a boolean singleton type of the argument |
+| `getunion(arg: {typelib})` | `typelib` | returns an immutable runtime representation of union type of its argument |
+| `getintersection(arg: {typelib})` | `typelib` | returns an immutable runtime representation of intersection type of its argument |
+| `newtable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?)` | `typelib` | returns a mutable runtime representation of a `table` type |
+| `newmetatable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?, metatable: typelib?)` | `typelib` | returns a mutable runtime representation of a metatable. `metatable` argument needs to be the same type as `typelib.newtable()` |
+| `newfunction(parameters: {typelib} \| typelib?, returns: {typelib} \| typelib?)` | `typelib` | returns a mutable runtime representation of a `function` type. Calling `newfunction(X)` will by default set `parameters` to `X` |
+| `type(arg: typelib)` | `string` | returns the tag of the argument ("nil", "unknown", "never", "any", "boolean", "number", "string", "boolean singleton", "string singleton", "negation", "union", "intersection", "table", "metatable", "function", "class") |
```

Comment:
Maybe we can even get rid of `"string singleton"` and `"boolean singleton"` and have both return just `"literal"`. Then you'd do this:

```luau
if ty:is("literal") and typeof(ty:getvalue()) == "string" then
  ...
end
```

Reply:
This did expose a problem with the current API design, type refinements on a function call doesn't do anything. So `typeof(ty:getvalue() == "string"` followed by another call to `ty:getvalue()` still has type `"string" | "boolean"`.

Extending type refinements to be able to refine the type of a function call will be difficult, so you end up having to push the value onto a register, and then refine that. :/

Reply:
I'd be okay with combining the singleton types into one tag, I think.

Reply:
I would be okay with that as well. In that case, I want to remove type.getstringsingleton / getbooleansingleton, and change it with: `type.getsingleton(arg: string | boolean)`